### PR TITLE
update scala version to 2.13

### PR DESCRIPTION
**Pull request by nlu90**
### Motivation

*Explain here the context, and why you're making that change. What is the problem you're trying to solve.*

### Modifications

- scala version from 2.12.17 to 2.13.12
- Pulsar version to 2.10.5
- Spark version to 3.4.1
- various scala code to clean warning message

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.

- [ ] This change is already covered by existing tests, such as:

- [ ] This change added tests and can be verified as follows:

### Documentation

Check the box below.

Need to update docs? 

- [ ] `doc-required`
- [x] `no-need-doc`
- [ ] `doc`

### CODE REVIEW

File: src/main/scala/org/apache/spark/sql/pulsar/PulsarHelper.scala [hunk 1]
```diff
@@ -288,10 +288,10 @@ private[pulsar] case class PulsarHelper(
       case Some((TopicSingle, value)) =>
         TopicName.get(value).toString :: Nil
       case Some((TopicMulti, value)) =>
-        value.split(",").map(_.trim).filter(_.nonEmpty).map(TopicName.get(_).toString)
+        value.split(",").toIndexedSeq.map(_.trim).filter(_.nonEmpty).map(TopicName.get(_).toString)
       case Some((TopicPattern, value)) =>
         getTopics(value)
-      case None =>
```
Comment (chaoqin-li1123) [inline src/main/scala/org/apache/spark/sql/pulsar/PulsarHelper.scala:294]: Why is this change needed?

Comment (nlu90) [reply src/main/scala/org/apache/spark/sql/pulsar/PulsarHelper.scala:294]: There's a Scala compiler warning says the case here is not exhaustive. 

The `_` refers to both `None` as well as other `Some(pattern)` cases that don't match the above 3 cases.

Comment (chaoqin-li1123) [reply src/main/scala/org/apache/spark/sql/pulsar/PulsarHelper.scala:294]: Thanks for the explanation, can we differentiate None and other illegal combination and have different error message for them?

Comment (nlu90) [reply src/main/scala/org/apache/spark/sql/pulsar/PulsarHelper.scala:294]: sure

**Review by chaoqin-li1123 — APPROVED** [review]
LGTM, thanks!

**Review by freeznet — APPROVED** [review]
